stdlib/public/core/DoubleWidth.swift.gyb

```diff
@@ -545,32 +492,34 @@ extension DoubleWidth: FixedWidthInteger {
   
   @_inlineable // FIXME(sil-serialize-all)
   public static func &<<=(lhs: inout DoubleWidth, rhs: DoubleWidth) {
-    // Need to use smart shifts here, since rhs can be > Base.bitWidth
-    let rhs = rhs & DoubleWidth(DoubleWidth.bitWidth - 1)
+    // We need to use smart shifts since rhs can be greater than Base.bitWidth.
+    let rhs = rhs & DoubleWidth(DoubleWidth.bitWidth &- 1)
 
+    // If Base.bitWidth >= 4, then rhs == High(rhs._storage.low).
     lhs._storage.high <<= High(rhs._storage.low)
 
     let lowInHigh = Base.bitWidth > rhs._storage.low
-      ? lhs._storage.low >> (numericCast(Base.bitWidth) - rhs._storage.low)
-      : lhs._storage.low << (rhs._storage.low - numericCast(Base.bitWidth))
+      ? lhs._storage.low >> (Low(Base.bitWidth) &- rhs._storage.low)
+      : lhs._storage.low << (rhs._storage.low &- Low(Base.bitWidth))
     lhs._storage.high |= High(truncatingIfNeeded: lowInHigh)
 
     lhs._storage.low <<= rhs._storage.low
   }
   
   @_inlineable // FIXME(sil-serialize-all)
   public static func &>>=(lhs: inout DoubleWidth, rhs: DoubleWidth) {
-    // Need to use smart shifts here, since rhs can be > Base.bitWidth
-    let rhs = rhs & DoubleWidth(DoubleWidth.bitWidth - 1)
+    // We need to use smart shifts since rhs can be greater than Base.bitWidth.
+    let rhs = rhs & DoubleWidth(DoubleWidth.bitWidth &- 1)
 
     lhs._storage.low >>= rhs._storage.low
 
+    // If Base.bitWidth >= 4, then rhs == High(rhs._storage.low).
     let highInLow = Base.bitWidth > rhs._storage.low
-      ? lhs._storage.high << (numericCast(Base.bitWidth) - rhs._storage.low)
-      : lhs._storage.high >> (rhs._storage.low - numericCast(Base.bitWidth))
+      ? lhs._storage.high << (High(Base.bitWidth) &- High(rhs._storage.low))
+      : lhs._storage.high >> (High(rhs._storage.low) &- High(Base.bitWidth))
     lhs._storage.low |= Low(truncatingIfNeeded: highInLow)
 
-    lhs._storage.high >>= High(truncatingIfNeeded: rhs._storage.low)
+    lhs._storage.high >>= High(rhs._storage.low)
```

Comment:
II think, if `High` is signed, this conversion can trap.

Reply:
It cannot (for any type where `Base.bitWidth >= 4`), because the value of `rhs` can be at most `DoubleWidth.bitWidth - 1`.

Reply:
...which is not to say that the implementation couldn't be improved. I've just taken the opportunity to streamline the logic and use masking shifts instead.